### PR TITLE
[OpenGL] fix NVIDIA's GLSL stuck at compile by sorting snodes

### PR DESCRIPTION
**Pull request by archibate**
Related issue = #804

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

### CODE REVIEW

**Comment by archibate** [comment]
Sorry for messing commits from #785 but I dep that and I base on some of that PR's codes to do this fix PR, please merge that first, thx!

**Comment by yuanming-hu** [comment]
Thanks, but I can only merge #785 in after all the comments there are resolved.

**Comment by archibate** [comment]
Thank for the reminder, it's resolved now!

**Comment by archibate** [comment]
Made git happy now.

**Comment by archibate** [comment]
Hello? PR title not good enough?

File: taichi/backends/opengl/opengl_api.cpp [hunk 1]
```diff
@@ -290,6 +292,20 @@ bool my_starts_with(std::string const &str, std::string const &pre) {
 }
 }  // namespace
 
+std::string display_kernel_info(std::string const &kernel_name,
+                                std::string const &kernel_source_code,
+                                int num_groups) {
+  if (!my_starts_with(kernel_name, "snode_") &&
```
Comment (yuanming-hu) [inline taichi/backends/opengl/opengl_api.cpp:298]: Please use `taichi::starts_with`

File: taichi/backends/opengl/opengl_api.cpp [hunk 2]
```diff
@@ -307,20 +323,13 @@ struct CompiledKernel {
                           int num_groups_,
                           RangeSizeEvaluator rse_,
                           const UsedFeature &used_)
-      : kernel_name(kernel_name_),
+      : kernel_name(
+            display_kernel_info(kernel_name_, kernel_source_code, num_groups_)),
```
Comment (yuanming-hu) [inline taichi/backends/opengl/opengl_api.cpp:327]: Please directly assign `kernel_name` and invoke `display_kernel_info` later.

File: taichi/struct/struct_opengl.cpp [hunk 3]
```diff
@@ -35,53 +34,36 @@ void OpenglStructCompiler::generate_types(const SNode &snode) {
...
-      // Is it possible for this to have no children?
-      stride_str = "0";
+    std::sort(smp.begin(), smp.end(),
+              [this](const std::pair<int, SNode *> &a,
+                     const std::pair<int, SNode *> &b) {
+                return stride_map_.at(a.second->node_type_name) <
+                       stride_map_.at(b.second->node_type_name);
+              });
+    for (auto &&[i, ch] : smp) {
+      // for (int i = 0; i < (int)snode.ch.size(); i++) {
```
Comment (yuanming-hu) [inline taichi/struct/struct_opengl.cpp:51]: Please remove this comment

File: taichi/struct/struct_opengl.cpp [hunk 4]
```diff
@@ -35,53 +34,36 @@ void OpenglStructCompiler::generate_types(const SNode &snode) {
...
-        gimme.push_back(stride_num);
-        stride_str += " + " + ch_node_name + "_stride";
-        stride_num += stride_map_[ch_node_name];
-      }
+      smp.push_back(std::pair<int, SNode *>(i, snode.ch[i].get()));
     }
-    if (stride_str.empty()) {
-      // Is it possible for this to have no children?
-      stride_str = "0";
+    std::sort(smp.begin(), smp.end(),
```
Comment (yuanming-hu) [inline taichi/struct/struct_opengl.cpp:44]: Please comment here to explain why a sort is necessary

**Pull request merged**